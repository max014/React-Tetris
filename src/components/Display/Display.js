import React, { Component } from 'react';
import { connect } from 'react-redux';
import Block from '../Block/Block';
import * as actionTypes from '../../store/actions/actionTypes';
import styles from './Display.module.css';

export class Display extends Component {
	constructor() {
	    super();
	    this.keydownHandler = this.keydownHandler.bind(this);
	}
	state = {
		time: Date.now()
	}

	keydownHandler(event) {
		this.props.setInput(event.keyCode)
	}

	componentDidMount() {
		// Controller
		document.addEventListener('keydown', this.keydownHandler);
		// Game Loop
		if(!this.state.paused){
			this.interval = setInterval(() => {
				this.props.update();
				this.setState({ time: Date.now()});
				// if game ends
				if(this.props.lost){
					this.props.end();
				}
			}, this.props.refreshRate);
		}
	}

	componentWillUnmount() {
		// Controller
		document.removeEventListener('keydown', this.keydownHandler);
		// Part of the loop
		clearInterval(this.interval);
	}

	render() {
		const inlineStyles = {
			height: this.props.height,
			width: this.props.step * 10
		}

		const board = this.props.board.map((row, y) => {
			return row.map((color, x) => {
				if (color !== 0){
					return (<Block x={x} y={y} color={color} key={x.toString() + "-" + y.toString()} />)
				}
				return null;
			})
		});

		return (
			<div className={styles.Display} style={inlineStyles}>
				{board}
			</div>
		);

	}
}

const mapStateToProps = state => {
    return {
        board: state.board,
        height: state.height,
        step: state.step,
        lost: state.lost
    };
}

const mapDispatchToProps = dispatch => {
    return {
        update: () => dispatch({type: actionTypes.UPDATE}),
        setInput: (input) => dispatch({type: actionTypes.SET_INPUT, input: input}),
        end: () => dispatch({type: actionTypes.END})
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Display);