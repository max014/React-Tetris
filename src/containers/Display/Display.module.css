.Display {
	position: relative;
	flex-grow: 0;
	flex-shrink: 0.1;
	background-color: #fff;
	margin: 0 25px;
}