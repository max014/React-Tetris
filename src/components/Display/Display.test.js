import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Display } from './Display';

describe('Display', () => {
    const update = jest.fn();
    const setInput = jest.fn();
    const end = jest.fn();
    const props = {
        board: [[0,0,0],[0,0,0,]],
        height: 800,
        step: 20,
        lost: false,
        refreshRate: 1000/30,
        update,
        setInput,
        end
    }
    jest.useFakeTimers();
    const display = shallow(<Display {...props}/>);
    it('renders properly', () => {
        expect(toJson(display)).toMatchSnapshot();
    });
    it('sets input', () => {
        display.instance().keydownHandler({ keyCode: 39 });
        expect(setInput).toHaveBeenCalled();
    });
    it('dispatches `update` from props', () => {
        jest.advanceTimersByTime(1000);
        expect(update).toHaveBeenCalled();
    });
    it('dispatches `end` when `lost` is true', () => {
        display.setProps({lost: true});
        jest.advanceTimersByTime(1000);
        expect(end).toHaveBeenCalled();
    });
    it('clears interval when it unmounts', () => {
        const interval = display.instance().interval;
        display.unmount();
        expect(clearInterval).toHaveBeenCalledWith(interval);
    });
});