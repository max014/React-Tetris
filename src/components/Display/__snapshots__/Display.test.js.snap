// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders properly 1`] = `
<div
  className="Display"
  style={
    Object {
      "height": 800,
      "width": 200,
    }
  }
/>
`;
